### PR TITLE
[PromotionAction] Configuring discounts in different currencies

### PR DESCRIPTION
**Pull request by GSadee**
| Q               | A
| --------------- | ---
| Bug fix?        | no |
| New feature?    | yes |
| BC breaks?      | yes |
| Related tickets |  |
| License         | MIT |

### CODE REVIEW

File: src/Sylius/Component/Core/Factory/PromotionActionFactory.php [hunk 1]
```diff
@@ -48,17 +48,17 @@ public function createNew()
     /**
      * {@inheritdoc}
      */
-    public function createFixedDiscount($amount)
+    public function createFixedDiscount($baseAmount, $amounts = array())
     {
-        return $this->createAction(FixedDiscountPromotionActionCommand::TYPE, ['amount' => $amount]);
+        return $this->createAction(FixedDiscountPromotionActionCommand::TYPE, ['base_amount' => $baseAmount]);
```
Comment (tuka217) [inline src/Sylius/Component/Core/Factory/PromotionActionFactory.php]: Why you pass here `$amounts`, since you do not use it in method?

File: features/promotion/managing_promotions/adding_promotion_with_action.feature [hunk 2]
```diff
@@ -10,8 +10,8 @@ Feature: Adding a new promotion with action
 
     @ui @javascript
     Scenario: Adding a new promotion with fixed discount
-        Given I want to create a new promotion
-        When I specify its code as "10_for_all_products"
+        When I want to create a new promotion
+        And I specify its code as "10_for_all_products"
```
Comment (michalmarcinkowski) [inline features/promotion/managing_promotions/adding_promotion_with_action.feature]: I would say that `Given` looks fine for me for all steps starting with "I want to". We open a page in this step, so it is like a preparation of an initial state 😉

Comment (GSadee) [inline features/promotion/managing_promotions/adding_promotion_with_action.feature]: So IMO these steps should looks like `Given I am creating a new promotion` or sth like that. Now we are extremely incoherent with naming steps like these and I guess that we should decide on some convention.

Comment (GSadee) [inline features/promotion/managing_promotions/adding_promotion_with_action.feature]: But of course I could revert this change if you want to 👍

File: features/promotion/managing_promotions/adding_promotion_with_action_in_different_currencies.feature [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    In order to give possibility to pay specifically less price for some goods
+    As an Administrator
+    I want to add a new promotion with action configured in different currencies to the registry
+
+    Background:
+        Given the store has currency "USD"
+        And the store has currency "GBP" with exchange rate 0.5
+        And the store operates on a channel named "Web-EU"
+        And that channel allows to shop using "USD" and "GBP" currencies
+        And that channel uses the "USD" currency by default
```
Comment (pamil) [inline features/promotion/managing_promotions/adding_promotion_with_action_in_different_currencies.feature]: `Web-EU` using `USD` by default, something seems wrong 💃

File: src/Sylius/Bundle/CoreBundle/spec/Handler/CartCurrencyChangeHandlerSpec.php [hunk 4]
```diff
@@ -58,21 +61,21 @@ function it_throws_handle_exception_when_a_cart_is_not_found(CartContextInterfac
 
     function it_handles_cart_currency_code_change(
         CartContextInterface $cartContext,
-        OrderUpdaterInterface $orderUpdater,
-        EventDispatcherInterface $eventDispatcher,
+        OrderUpdaterInterface $exchangeRateUpdater,
```
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/spec/Handler/CartCurrencyChangeHandlerSpec.php]: `$exchangeRateUpdater` -> `$orderExchangeRateUpdater` - it does not update the exchange rate, but only updates the order's exchange rate.

File: src/Sylius/Component/Core/spec/Promotion/Action/UnitFixedDiscountPromotionActionCommandSpec.php [hunk 5]
```diff
@@ -98,7 +109,63 @@ function it_applies_a_fixed_discount_on_every_unit_in_order(
...
+        AdjustmentInterface $promotionAdjustment1,
+        AdjustmentInterface $promotionAdjustment2,
+        Collection $originalItems,
+        Collection $units,
+        OrderInterface $order,
+        OrderItemInterface $orderItem1,
+        OrderItemUnitInterface $unit1,
+        OrderItemUnitInterface $unit2,
+        PromotionInterface $promotion,
+        CurrencyConverterInterface $currencyConverter
```
Comment (pamil) [inline src/Sylius/Component/Core/spec/Promotion/Action/UnitFixedDiscountPromotionActionCommandSpec.php]: Should be the 5th argument.

File: src/Sylius/Component/Core/spec/Promotion/Action/UnitFixedDiscountPromotionActionCommandSpec.php [hunk 6]
```diff
@@ -98,7 +109,63 @@ function it_applies_a_fixed_discount_on_every_unit_in_order(
...
+    ) {
+        $order->getCurrencyCode()->willReturn('PLN');
+
+        $currencyConverter->convertToBase(1000, 'PLN')->willReturn(250);
+
+        $order->getItems()->willReturn($originalItems);
```
Comment (pamil) [inline src/Sylius/Component/Core/spec/Promotion/Action/UnitFixedDiscountPromotionActionCommandSpec.php]: `$order->getItems()->willReturn(new ArrayCollection([$orderItem1]));` will let you delete `Collection $originalItems` mock.

File: src/Sylius/Component/Core/spec/Promotion/Action/UnitFixedDiscountPromotionActionCommandSpec.php [hunk 7]
```diff
@@ -98,7 +109,63 @@ function it_applies_a_fixed_discount_on_every_unit_in_order(
...
+        $originalItems->toArray()->willReturn([$orderItem1]);
+
+        $priceRangeFilter->filter([$orderItem1], ['base_amount' => 500, 'amounts' => ['PLN' => 1000]])->willReturn([$orderItem1]);
+        $taxonFilter->filter([$orderItem1], ['base_amount' => 500, 'amounts' => ['PLN' => 1000]])->willReturn([$orderItem1]);
+        $productFilter->filter([$orderItem1], ['base_amount' => 500, 'amounts' => ['PLN' => 1000]])->willReturn([$orderItem1]);
+
+        $orderItem1->getQuantity()->willReturn(2);
+        $orderItem1->getUnits()->willReturn($units);
```
Comment (pamil) [inline src/Sylius/Component/Core/spec/Promotion/Action/UnitFixedDiscountPromotionActionCommandSpec.php]: Same as for `$originalItems` collection.

File: features/promotion/managing_promotions/adding_promotion_with_action_in_different_currencies.feature (hunk 3 above)
Comment (lchrusciel) [inline features/promotion/managing_promotions/adding_promotion_with_action_in_different_currencies.feature]: Web-EU is using `GBP` and `USD`. In a few weeks/months/years/centuries/millenniums/eons(who knows?) both will be equally typical EU currency ;)

**Comment by pjedrzejewski** [comment]
Thank you Grzesiu!
